ui/client/components/graph/NodeDetailsContent.js

```diff
@@ -24,21 +26,61 @@ export class NodeDetailsContent extends React.Component {
   constructor(props) {
     super(props);
 
+    this.nodeObjectDetails = ProcessUtils.findNodeObjectTypeDefinition(this.props.node, this.props.processDefinitionData.processDefinition);
+
+    this.nodeDef = this.prepareNodeDef(props.node, this.nodeObjectDetails, props.processToDisplay)
+
     this.state = {
       ...TestRenderUtils.stateForSelectTestResults(null, this.props.testResults),
-      editedNode: _.cloneDeep(props.node),
+      editedNode: this.enrichNodeWithProcessDependentData(_.cloneDeep(props.node)),
       codeCompletionEnabled: true,
       testResultsToHide: new Set(),
     };
 
-    this.nodeObjectDetails = ProcessUtils.findNodeObjectTypeDefinition(this.props.node, this.props.processDefinitionData.processDefinition);
-
     let hasNoReturn = _.isNull(this.nodeObjectDetails.returnType)
     this.showOutputVar = hasNoReturn  === false || (hasNoReturn === true && this.state.editedNode.outputVar)
 
     this.generateUUID("fields");
   }
 
+  prepareNodeDef(node, nodeObjectDetails, processToDisplay) {
+    if (NodeUtils.nodeType(node) === "Join") {
+      return new JoinDef(node, nodeObjectDetails, processToDisplay)
+    } else {
+      return null
+    }
+  }
+
+  // should it be here or somewhere else (in the reducer?)
+  enrichNodeWithProcessDependentData(node) {
```

Comment:
or maybe it should go to BranchParameters?

Reply:
I'm not convinced about it. For now, we hold state of node in `NodeDetailsContent`, nested components only call callbacks changing this state by `setNodeDataAt` - see `ExpressionSuggest`, `ExpressionWithFixedValues`, `MapVariable`, `Fields`. It would be not consistent with this behavious. Also it would generate unnecessary initial state change.

Reply:
hmm... ok, let's leave it till refactoring ;)